README.md

```diff
@@ -0,0 +1,453 @@
+- **queryJSON:** will use a `Setoid` instance that performs a strict equality check after transforming the data via JSON stringification to compare inputs.
+
+Some examples will help clarify:
+```ts
+/*
+  this implementation will always re-run the `Fetch` function
+  even if valid cached data is already present
+  and use shallow equality to compare input output when called
+*/
+const myQuery = queryShallow(fetchFunction, refetch);
+
+/*
+  this implementation will never run the `Fetch` function
+  unless no valid data is present in the Cache
+  and use strict equality to compare input/output when called
+*/
+const myQuery = queryStrict(fetchFunction, available);
+
+/*
+  this implementation will run the `Fetch` function only if no valid data is present in the Cache
+  or t > 10000 ms passed till the last time data was fetched
+  and use strict equality to compare input output when called
+*/
+const myQuery = queryJSON(fetchFunction, expire(10000));
+```
+
+Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as a `key` to store the result obtained:
+```
+// usersCache is empty
+usersCache: {}
+
+//a user is fetched
+getUser({ userId: 1 }) -> { userName: "Mario" }
+
+// usersCache is now populated
+usersCache: {
+  [{ userId: 1 }]: { userName: Mario }
+}
+```
+
+From that moment onwards, when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will:
+
+1. attempt to retrive data from the [**`Cache`**](#Cache)
+2. match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).
+
+If the result is valid it is returned, otherwise the `Fetch` function will be re-run in order to get new and valid data. The two flows are relatively simple:
+##### Valid CacheValue
+!["cached flow"](docs/CachedValue.svg)
+
+##### Invalid CacheValue
+when you call `run` or `subscribe` on a `query` with a combination of `inputs` that was never used before (or whose last use ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
+!["cached flow"](docs/UncachedOrErrorValue.svg)
+
+There are two ways to get a query result:
+
+```ts
+type Error = 'there was an error';
+type User = { userName: String };
+
+declare function getUser(userId: number): TaskEither<Error, User>
+
+const userQuery: ObservableQuery<number, Error, User> = query(
+  getUser,
+  refetch
+);
+
+declare function dispatchError(e: Error): void;
+declare function setCurrentUser(e: User): void;
+
+// feeding your query to `observe` will give you an observable on the query
+// N.B. until now no fetch is yet attempted, avenger will wait until the first subscription is issued
+const observable: Observable<QueryResult<Error, User>> = observe(userQuery);
+
+// after this the fetch function triggers
+observable.subscribe(dispatchError, setCurrentUser);
```

Comment:
What if I want two subscribers? Does the fetch happen twice according to cache semantics?

Reply:
exactly, it depends on the strategy passed to the `query` util